src/embeddedContentFileSystemProvider.ts

```diff
@@ -0,0 +1,328 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All Rights Reserved.
+ */
+
+import * as vscode from "vscode";
+import * as sarif from "sarif";
+import * as fs from "fs";
+import * as path from "path";
+import { BinaryContentRenderer } from "./binaryContentRenderer";
+import { LogReader } from "./logReader";
+
+/**
+ * Represents information parsed from the 'sarifEmbeddedContent' URI.
+ * Returned from  @see EmbeddedContentFileSystemProvider.parseUri
+ */
+interface ParsedUriData {
+    /**
+     * The SARIF log the URI was created from.
+     */
+    readonly log: vscode.Uri;
+
+    /**
+     * The original file name specified in the location object of the SARIF.
+     */
+    readonly artifactUri: vscode.Uri;
+
+    /**
+     * The run index in the SARIF log.
+     */
+    readonly runIndex: number;
+
+    /**
+     * The artifact index in the SARIF log (relative to the run).
+     */
+    readonly artifactIndex: number;
+}
+
+/**
+ * Represents information parsed from the 'sarifEmbeddedContent' URI.
+ * Used to encode and decode URI information in  @see EmbeddedContentFileSystemProvider.parseUri
+ * and @see EmbeddedContentFileSystemProvider.tryCreateUri
+ */
+interface EncodedUriData {
+    /**
+     * The SARIF log the URI was created from.
+     */
+    readonly log: string;
+
+    /**
+     * The original file name specified in the location object of the SARIF.
+     */
+    readonly artifactUri: string;
+
+    /**
+     * The run index in the SARIF log.
+     */
+    readonly runIndex: number;
+
+    /**
+     * The artifact index in the SARIF log (relative to the run).
+     */
+    readonly artifactIndex: number;
+}
+
+/**
+ * Use to return the artifact contents and the data parsed from the embedded artifact URI.
+ * @see EmbeddedContentFileSystemProvider.getArtifactInformationFromUri
+ */
+interface ArtifactInformationFromUri {
+    /**
+     * The artifact.
+     */
+    artifact: sarif.Artifact;
+
+    /**
+     * The artifact contents.
+     */
+    contents: sarif.ArtifactContent;
+
+    /**
+     * The information parsed from the embedded URI.
+     */
+    parsedUriData: ParsedUriData;
+
+    /**
+     * The SARIF log file.
+     */
+    log: sarif.Log;
+}
+
+/**
+ * A file system provider that handles embedded content.
+ */
+export class EmbeddedContentFileSystemProvider implements vscode.FileSystemProvider, vscode.Disposable {
+    private disposables: vscode.Disposable[] = [];
+
+    // The URI scheme for the content.
+    private static EmbeddedContentScheme: string = 'sarifEmbeddedContent';
```

Comment:
>sarifEmbeddedContent [](start = 52, length = 20)

We should really rename this to sarifArtifactContent because the term "embedded" is not used in the spec.

Reply:
Agreed.